Conformance test for GRPCRouteRequestMirror #3514

TaranpreetNatt opened the pull request:
**What type of PR is this?**
/kind test
/area conformance-test

**What this PR does / why we need it**:
Adds conformance tests for GRPCRoute request mirroring functionality to enable gradual implementation support across Gateway API implementations.

**Which issue(s) this PR fixes**:
Fixes #3514

**Does this PR introduce a user-facing change?**:

    NONE

TaranpreetNatt committed:
Added conformance test for GRPCRouteRequestMirror, incomplete testing logic, will add in later commit

TaranpreetNatt committed:
Added logic to check if the headers were modified correctly by the GRPCRoute

TaranpreetNatt committed:
Finished GRPCRouteRequestMirror conformance test

TaranpreetNatt committed:
Improved variable naming: receivedMap -> receivedHeadersMap for clarity

LiorLieberman commented:
/ok-to-test
/cc

LiorLieberman reviewed (commented):
Thanks!

TaranpreetNatt commented on conformance/utils/grpc/mirror.go:
```diff
@@ -0,0 +1,73 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func ExpectMirroredRequest(t *testing.T, client client.Client, clientset clientset.Interface, mirrorPods []MirroredBackend) {
```
We can use that function, I would need to refactor it as the regex the function is looking for is not the same. But it is very doable. Would you like me to do that in this PR? 

For context: it's this line

For http: `mirrorLogRegexp := regexp.MustCompile(fmt.Sprintf("Echoing back request made to \\%s to client", path))`

For grpc:  `mirrorLogRegexp := regexp.MustCompile("Received over plaintext:")`

TaranpreetNatt replied:
I refactored the http ExpectMirroredRequest to be more generic and used it in the grpc mirror tests. I also deleted the duplicated function and file.

LiorLieberman commented:
now I see that we already have a PR open for that -- https://github.com/kubernetes-sigs/gateway-api/pull/4013

TaranpreetNatt committed:
Refactored ExprectMirroredRequest to one function and be shared between HTTP and GRPC

Makes the HTTP mirror validation function configurable with a logPattern
parameter so it can be reused by gRPC tests, eliminating code duplication.

  - Parameterize logPattern in HTTP ExpectMirroredRequest
  - Update gRPC to use shared function with gRPC-specific pattern
  - Remove duplicate grpc/mirror.go implementation
  - Make error messages protocol-agnostic

  Addresses reviewer feedback in PR #4015.

TaranpreetNatt commented on conformance/tests/grpcroute-request-mirror.go:
```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2023 The Kubernetes Authors.
```
I changed it to 2025

TaranpreetNatt committed:
Fixed linting errors

TaranpreetNatt commented on conformance/tests/grpcroute-request-mirror.go:
Sure, I can look into it

TaranpreetNatt commented:
> now I see that we already have a PR open for that -- #4013

How does it work when two PRs are open for the same issue?

TaranpreetNatt committed:
Merge branch 'main' into tests/add-grpcRoute-coverage-in-request-mirror-tests

TaranpreetNatt committed:
Fixing merge conflicts from main

TaranpreetNatt committed:
Refactor ExpectMirroredRequest to accomodate for http and grpc differences

- Add http.ExpectMirroredRequestWithPattern() with shared validation
logic
- Both protocols now share the same core validation: goroutine spawning,
log reading, regex matching, and error handling

TaranpreetNatt committed:
Fix linting errors

LiorLieberman reviewed (commented):
Also, just tested locally with istio and it fails

    grpc.go:333: 2026-01-23T19:13:20.719413473Z: Response expectation failed for request: {<nil>  <nil> 0xc000eba258 {OK 0xc0000aaf48 <nil> <nil> [X-Header-Remove]} grpc-infra-backend-v1 gateway-conformance-infra [{{grpc-infra-backend-v2 gateway-conformance-infra} <nil>}] }  not ready yet: expected pod name to start with grpc-infra-backend-v1, got grpc-infra-backend-v2-5d756c89dd-rtlbv (after 29.085549379s)
    
 will be great if you can fix it, run it and report back

TaranpreetNatt committed:
Fix typo in logPattern

TaranpreetNatt committed:
Add mesh grpcroute-request-mirror

TaranpreetNatt committed:
Fix dates

TaranpreetNatt commented on conformance/tests/grpcroute-request-mirror.go:
```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2025 The Kubernetes Authors.
```
Fixed

TaranpreetNatt committed:
Fix linter issue

TaranpreetNatt commented:
> Also, just tested locally with istio and it fails
> 
> 
> 
>     grpc.go:333: 2026-01-23T19:13:20.719413473Z: Response expectation failed for request: {<nil>  <nil> 0xc000eba258 {OK 0xc0000aaf48 <nil> <nil> [X-Header-Remove]} grpc-infra-backend-v1 gateway-conformance-infra [{{grpc-infra-backend-v2 gateway-conformance-infra} <nil>}] }  not ready yet: expected pod name to start with grpc-infra-backend-v1, got grpc-infra-backend-v2-5d756c89dd-rtlbv (after 29.085549379s)
> 
>     
> 
>  will be great if you can fix it, run it and report back

I ran the test using Istio locally and mine passed. Could you please try again.

TaranpreetNatt commented:
@LiorLieberman Can I please get a review for this PR.

rikatz commented:
Please rebase this test, and if possible ping me or some conformance reviewer on #sig-network-gateway-api on Slack

Thanks!
